.github/workflows/ci.yml

```diff
@@ -22,33 +20,99 @@ jobs:
       # and hyphens.
       run: echo "TAG=branch-${GITHUB_REF_NAME//[^A-Za-z0-9._-]/-}" | tee -a $GITHUB_ENV
 
-    - run: ./devel/start-localhost-registry $REGISTRY_PORT
+    # GITHUB_TOKEN is unreliable¹ so use a token from nextstrain-bot.
+    # ¹ https://github.com/docker/build-push-action/issues/463#issuecomment-939394233
+    - uses: docker/login-action@v1
+      with:
+        registry: ghcr.io
+        username: nextstrain-bot
+        password: ${{ secrets.GH_TOKEN_NEXTSTRAIN_BOT_MANAGE_PACKAGES }}
 
-    - run: ./devel/build -r localhost:$REGISTRY_PORT -t $TAG
+    - run: ./devel/build -r ghcr.io -t $TAG
 
-    - if: github.event_name != 'pull_request'
-      uses: docker/login-action@v1
-      with:
-        username: ${{ secrets.DOCKER_LOGIN }}
-        password: ${{ secrets.DOCKER_PASSWORD }}
+    outputs:
+      tag: ${{ env.TAG }}
+
+  test:
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
 
-    - if: github.event_name != 'pull_request' && startsWith(env.TAG, 'branch-')
-      name: Copy $TAG images to Docker Hub (non-default branch)
-      run: ./devel/copy-images -i localhost:$REGISTRY_PORT -o docker.io -t $TAG
+    - uses: docker/login-action@v1
+      with:
+        registry: ghcr.io
+        username: nextstrain-bot
+        password: ${{ secrets.GH_TOKEN_NEXTSTRAIN_BOT_MANAGE_PACKAGES }}
 
     - uses: actions/setup-python@v4
       with:
         python-version: ~3
+
     - name: Run zika-tutorial
       run: |
         git clone https://github.com/nextstrain/zika-tutorial
         pip install nextstrain-cli
-        nextstrain build --image localhost:$REGISTRY_PORT/nextstrain/base:$TAG zika-tutorial -F
+        nextstrain build --image ghcr.io/nextstrain/base:${{ needs.build.outputs.tag }} zika-tutorial -F
 
-    - if: startsWith(env.TAG, 'build-')
-      name: Copy $TAG + latest images to Docker Hub (default branch)
+  push-branch:
+    if: startsWith(needs.build.outputs.tag, 'branch-') && github.event_name != 'pull_request'
+    needs: build
```

Comment:
I didn't think of ways, but thinking now and it seems difficult because only `push-branch` should run before `test`.

Having them separate provides a more intuitive UI page for the workflow run:

![CI page](https://user-images.githubusercontent.com/13424970/199307686-423d010d-aeca-4212-b96d-b19860f5259d.png)

Reply:
Nod. I _think_ we could make a single `push` job with a dynamically-set `needs:` and thus avoid the "one of the push-* jobs never runs" behaviour, but it's just fine as-is with two jobs too.

Reply:
I played with this a bit and don't think it's possible to dynamically set `needs`: https://github.com/community/community/discussions/38492